### PR TITLE
Add imptests for delaying actions enactment

### LINKED ISSUE #4127: Imp test for checking delaying works as expected
Certain proposal type prevent others from being ratified due to the feature of delaying. We need to write a few unit tests that ensure this indeed happens as expected. Note that there is a corner case of an expiring proposal being delayed (see #4109), that scenario should not be considered in this ticket.

Some scenarios that come to mind:
1. Proposal that triggers a delay prevents another dependent proposal from being enacted until the subsequent epoch, despite both of them having enough votes
2. Proposal delays enactment of all other types of proposals until the next epoch, despite them having enough votes

### PR DESCRIPTION
**Pull request by aniketd**
# Description

Resolves #4127 

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/RatifySpec.hs [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+      getLastEnactedParameterChange `shouldReturn` SNothing
+      passEpoch
+      -- all three actions, pGai0, pGai1 and pGai2, are enacted one
+      -- after the other in the same epoch
+      getLastEnactedParameterChange `shouldReturn` SJust (GovPurposeId pGai2)
+      getParameterChangeProposals `shouldReturn` Map.empty
+    describe "An action expires when delayed enough even after being ratified" $ do
+      it "Same lineage" $ do
+        (dRep, committeeMember, _gpi) <- electBasicCommittee
+        conwayModifyPParams $ ppGovActionLifetimeL .~ EpochInterval 2
```
Comment (lehins) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/RatifySpec.hs]: YOU should be using the type class method instead, also I think this one is redundant, since `electBasicCommittee` already sets the same value
```suggestion
        modifyPParams $ ppGovActionLifetimeL .~ EpochInterval 2
```

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/RatifySpec.hs [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        passEpoch
+        getLastEnactedConstitution `shouldReturn` SJust (GovPurposeId gai1)
+        passEpoch
+        getLastEnactedConstitution `shouldReturn` SJust (GovPurposeId gai2)
+        getConstitutionProposals `shouldReturn` Map.empty
+        passEpoch
+        getLastEnactedConstitution `shouldReturn` SJust (GovPurposeId gai2)
+      it "Other lineage" $ do
+        (dRep, committeeMember, _gpi) <- electBasicCommittee
+        conwayModifyPParams $ ppGovActionLifetimeL .~ EpochInterval 2
```
Comment (lehins) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/RatifySpec.hs]: Ditto

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/RatifySpec.hs [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+      passEpoch
+      getLastEnactedConstitution `shouldReturn` SJust (GovPurposeId gai1)
+      passEpoch
+      getLastEnactedConstitution `shouldReturn` SJust (GovPurposeId gai2)
+      passEpoch
+      getLastEnactedConstitution `shouldReturn` SJust (GovPurposeId gai3)
+      getConstitutionProposals `shouldReturn` Map.empty
+    it "A delaying action delays all other actions even when all of them may be ratified in the same epoch" $ do
+      (dRep, committeeMember, _gpi) <- electBasicCommittee
+      void $ conwayModifyPParams $ ppGovActionLifetimeL .~ EpochInterval 5
```
Comment (lehins) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/RatifySpec.hs]: What's up with `void`? :grin: It returns `()`
```suggestion
      modifyPParams $ ppGovActionLifetimeL .~ EpochInterval 5
```

**Review by lehins — APPROVED** [review]
This looks great!

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/RatifySpec.hs (hunk 3 above)
Comment (aniketd) [reply eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/RatifySpec.hs]: Oh that's a left over :sweat: :pray:

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/RatifySpec.hs (hunk 1 above)
Comment (aniketd) [reply eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/RatifySpec.hs]: I thought I would keep it explicit, but I realise after thinking more that it makes little sense to do that, :+1:
